### ISSUE TITLE
PacketPlayInWindowClick crashes the server

### ISSUE THREAD

**Issue by Xayanix** [PaperSpigot/Paper-1.7#58]
[ProtocolLib] Received PacketPlayInWindowClick[14, legacy: 102] from <nickname>

Spamming this packets cause massive cpu load.

**Comment by caoli5288** [comment]
Maybe ProtocolLib's issue?

**Comment by Xayanix** [comment]
No, lol.

I used ProtocolLib for check which packet is he spamming only.

**Comment by Zbob750** [comment]
Will eventually get around to investigating this and potentially adding some sort of rate limiting.
Whether or not 1.7 sees a fix will depend on how severe we find the issue afterwards.

**Comment by caoli5288** [comment]
@Xayanix Ohh, I have opened a similar issue. #46

**Comment by Xayanix** [comment]
PacketPlayInWindowClick throotling fixes issue.

W dniu czwartek, 8 października 2015 Zhang <notifications@github.com>
napisał(a):

**Comment by MrPowerGamerBR** [comment]
@Xayanix Did you find a solution?

There is a private client (Unfortunately I don't have it) which can crash servers with this packet.

**Comment by Xayanix** [comment]
Yes, download ProtocoLib and use /packet for listen which packet is he sending.
Then make own plugin for filtering/throttling these packets and its all.

**Comment by MrPowerGamerBR** [comment]
@Xayanix thanks! They aren't flooding and crashing my Server (yet.), however, it is known that a "menu gui" exploit can crash Servers, so it is better fix it before they attack :+1:

**Comment by MrPowerGamerBR** [comment]
@Xayanix Maybe maybe maybe I'm doing something wrong, because I don't know how to filter those packets:

Here's my code:
```
public HashMap<String, Long> listOfThrottlers = new HashMap<String, Long>();
	
	@Override
	public void onEnable() {
		ProtocolLibrary.getProtocolManager().addPacketListener(new PacketAdapter(this, new PacketType[] { Play.Client.WINDOW_CLICK }) {
			public void onPacketReceiving(final PacketEvent event) {
				
				if (listOfThrottlers.containsKey(event.getPlayer().getAddress().getHostString())) {
					Long l = listOfThrottlers.get(event.getPlayer().getAddress().getHostString());
					
					Long diff = System.currentTimeMillis() - l;
					
					Long inSec = diff / 1000;
					
					if (inSec >= 0.5) {
						event.setCancelled(true);
					}
				}
				listOfThrottlers.put(event.getPlayer().getAddress().getHostString(), System.currentTimeMillis());
			}
		});
	}
```
However, is filtering just canelling the event? Thanks!

**Comment by Xayanix** [comment]
Yes.
You can kick the player optionally.

**Comment by Xayanix** [comment]
You should allow player to use maxmimum 5 clicks per second.

**Comment by MrPowerGamerBR** [comment]
@Xayanix  :+1:

**Comment by MrPowerGamerBR** [comment]
Another question, what does the 'limit-player-interactions' option does?

In the PaperSpigot features pages says that
"Prevent players from spamming interaction packets
Prevents malicious players from hurting the server, but may cause issues for players with a high ping"

Shouldn't this block the PacketPlayInWindowClick spam? Or it is related to another packet spam?

**Comment by Xayanix** [comment]
I think its bugged, i set for ex. max-packet-per-player to 0 and i was still able to crash my server.

**Comment by madtomic** [comment]
Will this plugin fix this problem?

http://dev.bukkit.org/bukkit-plugins/clicks-per-second/

**Comment by HunterGPlays** [comment]
Isn't that for PvP, not for GUIs?

-----Original Message-----

**Comment by Xayanix** [comment]
@madtomic 
No way.
